### PR TITLE
Adding optional web hook part

### PR DESCRIPTION
**Pull request by mnkg561**
close https://github.com/keikoproj/iam-manager/issues/41

**Could you share the solution in high level?**
Following are the list of tasks taken care in this PR..
1. Webhook will not be registered by default and can be managed it by a config map param
2. Handled maximum number of roles validation in the controller itself
3. Added kustomize files for non-webhook and updated makefile
4. IAM Role name will be derived from the name of the resource with the prefix k8s_
5. Max number of roles per namespace can be managed by config map and default is 1
6. Validate IAM Policy action and resource is also moved to controller.

**Could you share the test results?**
Installation without web hook observe that no web hook registration is happening while starting up
```
mtvl15367e28a:~ nmogulla$ k logs -f -l app=iam-manager
2020-02-22T02:54:42.287Z	INFO	internal.config.properties.RunConfigMapInformer	Starting config map informer
2020-02-22T02:54:42.905Z	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": ":8080"}
2020-02-22T02:54:42.905Z	DEBUG	main.setup	Setting up reconciler with manager
2020-02-22T02:54:42.984Z	DEBUG	k8s.client.SetUpEventHandler	Successfully added event broadcaster
2020-02-22T02:54:42.985Z	INFO	controller-runtime.controller	Starting EventSource	{"controller": "iamrole", "source": "kind source: /, Kind="}
2020-02-22T02:54:42.985Z	INFO	v1alpha1.LoadProperties	loading k8s client
2020-02-22T02:54:42.986Z	INFO	main.setup	Registering controller
2020-02-22T02:54:42.987Z	INFO	controller-runtime.manager	starting metrics server	{"path": "/metrics"}
2020-02-22T02:54:59.171Z	INFO	controller-runtime.controller	Starting Controller	{"controller": "iamrole"}
```

New custom role creation:
```
2020-02-22T03:14:49.321Z	INFO	controllers.iamrole_controller.Reconcile	Start of the request	{"request_id": "56fc2bd0-972d-44b2-9fbb-74973b61d6e5"}
2020-02-22T03:14:49.321Z	INFO	controllers.iamrole_controller.HandleReconcile	state of the custom resource 	{"request_id": "56fc2bd0-972d-44b2-9fbb-74973b61d6e5", "state": "Error"}
2020-02-22T03:14:49.321Z	DEBUG	controllers.iamrole_controller.HandleReconcile	Going to requeue it as part of exponential back off after this try	{"request_id": "56fc2bd0-972d-44b2-9fbb-74973b61d6e5", "count": 2, "time in ms": 400}
2020-02-22T03:14:49.321Z	DEBUG	awsapi.iam.CreateRole	Initiating api call	{"request_id": "56fc2bd0-972d-44b2-9fbb-74973b61d6e5"}
2020-02-22T03:14:49.489Z	DEBUG	awsapi.iam.CreateRole	Attaching Tag	{"request_id": "56fc2bd0-972d-44b2-9fbb-74973b61d6e5"}
2020-02-22T03:14:49.490Z	DEBUG	awsapi.iam.TagRole	Initiating api call	{"request_id": "56fc2bd0-972d-44b2-9fbb-74973b61d6e5"}
2020-02-22T03:14:49.623Z	DEBUG	awsapi.iam.TagRole	Successfully completed TagRole call	{"request_id": "56fc2bd0-972d-44b2-9fbb-74973b61d6e5"}
2020-02-22T03:14:49.623Z	DEBUG	awsapi.iam.CreateRole	Attaching Permission Boundary	{"request_id": "56fc2bd0-972d-44b2-9fbb-74973b61d6e5"}
2020-02-22T03:14:49.623Z	DEBUG	awsapi.iam.AddPermissionBoundary	Initiating api call	{"request_id": "56fc2bd0-972d-44b2-9fbb-74973b61d6e5"}
2020-02-22T03:14:49.721Z	DEBUG	awsapi.iam.AddPermissionBoundary	Successfuly added permission boundary	{"request_id": "56fc2bd0-972d-44b2-9fbb-74973b61d6e5"}
2020-02-22T03:14:49.721Z	DEBUG	awsapi.iam.CreateRole	Attaching Managed policies	{"request_id": "56fc2bd0-972d-44b2-9fbb-74973b61d6e5"}
2020-02-22T03:14:49.721Z	DEBUG	awsapi.iam.AttachManagedRolePolicy	Initiating api call	{"request_id": "56fc2bd0-972d-44b2-9fbb-74973b61d6e5"}
2020-02-22T03:14:49.839Z	DEBUG	awsapi.iam.AttachManagedRolePolicy	Successfully attached managed policies	{"request_id": "56fc2bd0-972d-44b2-9fbb-74973b61d6e5"}
2020-02-22T03:14:49.839Z	DEBUG	awsapi.iam.CreateRole	Attaching Inline role policies	{"request_id": "56fc2bd0-972d-44b2-9fbb-74973b61d6e5"}
2020-02-22T03:14:49.839Z	DEBUG	awsapi.iam.AttachInlineRolePolicy	Initiating api call	{"request_id": "56fc2bd0-972d-44b2-9fbb-74973b61d6e5"}
2020-02-22T03:14:49.967Z	DEBUG	awsapi.iam.AttachInlineRolePolicy	Successfully completed attaching InlineRolePolicy	{"request_id": "56fc2bd0-972d-44b2-9fbb-74973b61d6e5"}
2020-02-22T03:14:49.972Z	DEBUG	controller-runtime.controller	Successfully Reconciled	{"controller": "iamrole", "request": "dev-patterns-iksmanager-usw2-ppd-idev/dev-patterns-iksmanager-usw2-ppd-idev"}
```

### CODE REVIEW

File: internal/config/properties.go [hunk 1]
```diff
@@ -80,48 +84,72 @@ func LoadProperties(env string, cm ...*v1.ConfigMap) error {
...
+		allowedPolicyAction:       allowedPolicyAction,
+		restrictedPolicyResources: restrictedPolicyResources,
+		restrictedS3Resources:     restrictedS3Resources,
+		awsMasterRole:             awsMasterRole,
+	}
+
+	//Defaults
+	isWebhook := cm[0].Data[propertWebhookEnabled]
+	if isWebhook != "" {
+		Props.isWebhookEnabled = isWebhook
```
Comment (kshamajain99) [inline internal/config/properties.go]: Why don't we change logic to 
```
if isWebhook == "true" {
     Props.isWebhookEnabled = true
} else {
     Props.isWebhookEnabled = false
}
```
where isWebhookEnabled is a boolean 

This way we cover both cases of 
1. No value for webhook.enable is set in config-map then default is false
2. By mistake user entering junk value like `yes`, `xyz`

Also, now we can simplify below function
```
func (p *Properties) IsWebHookEnabled() bool {
	resp := false
	if p.isWebhookEnabled == "true" {
		resp = true
	}
	return resp
}
```
to 
```
func (p *Properties) IsWebHookEnabled() bool {
	return p.isWebHookEnabled
}
```

File: internal/config/properties.go [hunk 2]
```diff
@@ -80,48 +84,72 @@ func LoadProperties(env string, cm ...*v1.ConfigMap) error {
...
 	}
 
-	allowedPolicyAction := strings.Split(cm[0].Data[propertyIamPolicyWhitelist], separator)
-	restrictedPolicyResources := strings.Split(cm[0].Data[propertyIamPolicyBlacklist], separator)
-	restrictedS3Resources := strings.Split(cm[0].Data[propertyIamPolicyS3Restricted], separator)
-	awsMasterRole := cm[0].Data[propertyAwsMasterRole]
+	managedPermissionBoundaryPolicy := cm[0].Data[propertyPermissionBoundary]
+
+	if managedPermissionBoundaryPolicy == "" {
+		managedPermissionBoundaryPolicy = "k8s-iam-manager-cluster-permission-boundary"
```
Comment (kshamajain99) [inline internal/config/properties.go]: shouldn't we make default permission boundary name to be unique per cluster like `k8s-iam-manager-{cluster-name}-permission-boundary`

File: internal/config/properties.go [hunk 3]
```diff
@@ -80,48 +84,72 @@ func LoadProperties(env string, cm ...*v1.ConfigMap) error {
...
+	}
+
+	Props.managedPermissionBoundaryPolicy = fmt.Sprintf(PolicyARNFormat, awsAccountID, managedPermissionBoundaryPolicy)
```
Comment (kshamajain99) [inline internal/config/properties.go]: Why did we remove this check here?
```
if !strings.HasPrefix(managedPermissionBoundaryPolicy, "arn:aws:iam::") {
		managedPermissionBoundaryPolicy = fmt.Sprintf(PolicyARNFormat, awsAccountID, managedPermissionBoundaryPolicy)
	}
```

File: internal/config/properties.go (hunk 1 above)
Comment (mnkg561) [reply internal/config/properties.go]: Fair. i will change it

File: internal/config/properties.go (hunk 2 above)
Comment (mnkg561) [reply internal/config/properties.go]: No. By default, we don't need to restrict users to have cluster name in the permission boundary. There will be an issue only if 2 clusters in the same account. So, we consider 2 clusters usecase as corner use case but default usecase is single cluster per account. I might need to update the default param value for CFN if not done

File: internal/config/properties.go (hunk 3 above)
Comment (mnkg561) [reply internal/config/properties.go]: updated

**Pull request merged**